### ISSUE TITLE
Missed factorization investigation

### ISSUE THREAD

**Issue by jkwashbourne** [devitocodes/devito#1290]
A prototype SSA VTI op sees a factor >5 in flops per output point based on whether a common term is factored in the equation or not. Example code below, two cases to run:

Case 1: lines 43,44 active, line 45 commented:
```
eta = Function(name='eta', grid=v.grid, space_order=v.space_order)
eta.data[:] = np.sqrt((2.0 * (e.data - d.data)) / (f + 2.0 * e.data))
# eta = sqrt((2.0 * (e - d)) / (f + 2.0 * e))
```

Case 2: lines 43,44 commented, line 45 active
```
# eta = Function(name='eta', grid=v.grid, space_order=v.space_order)
# eta.data[:] = np.sqrt((2.0 * (e.data - d.data)) / (f + 2.0 * e.data))
eta = sqrt((2.0 * (e - d)) / (f + 2.0 * e))
```

Run 1, with case 1:
```
env DEVITO_LANGUAGE="openmp" DEVITO_LOGGING="DEBUG" OMP_NUM_THREADS=32  python3 issue.py

Flops reduction after symbolic optimization: [701 --> 169]
```

Run 2, with case 2:
```
env DEVITO_LANGUAGE="openmp" DEVITO_LOGGING="DEBUG" OMP_NUM_THREADS=32  python3 issue.py

Flops reduction after symbolic optimization: [861 --> 532]
```

issue.py copied below
```
import numpy as np
from sympy import sqrt

from devito import Function, TimeFunction, Eq, Operator, info
from examples.seismic import RickerSource
from examples.seismic.skew_self_adjoint import (default_setup_iso,
                                                setup_w_over_q)

# Default constants
space_order = 8
shape = (601, 601, 301)
dtype = np.float32
npad = 20
qmin = 0.1
qmax = 1000.0
tmax = 500.0
fpeak = 0.010
omega = 2.0 * np.pi * fpeak
vp = 1.5
vs = 0.6
f = 1 - vs**2 / vp**2

# Define the model
b, v, time_axis, src_coords, rec_coords = \
    default_setup_iso(npad, shape, dtype, tmax=tmax)
wOverQ = Function(name='wOverQ', grid=v.grid, space_order=v.space_order)
setup_w_over_q(wOverQ, omega, qmin, qmax, npad)
e = Function(name='e', grid=v.grid, space_order=v.space_order)
d = Function(name='d', grid=v.grid, space_order=v.space_order)
v.data[:] = 1.5
e.data[:] = 0.2
d.data[:] = 0.1

# State variable
p = TimeFunction(name='p', grid=v.grid, time_order=2, space_order=space_order)
m = TimeFunction(name='m', grid=v.grid, time_order=2, space_order=space_order)

print(p.dimensions)

# Symbols for dimension
t, x, y, z = p.dimensions

# Eta 
# eta = Function(name='eta', grid=v.grid, space_order=v.space_order)
# eta.data[:] = np.sqrt((2.0 * (e.data - d.data)) / (f + 2.0 * e.data))
eta = sqrt((2.0 * (e - d)) / (f + 2.0 * e))

# Define the spatial derivative operators
# We use g, g_tilde here because the TTI operator will use the same notation
def g1(field):
    return field.dx(x0=x+x.spacing/2)

def g1_tilde(field):
    return field.dx(x0=x-x.spacing/2)

def g2(field):
    return field.dy(x0=y+y.spacing/2)

def g2_tilde(field):
    return field.dy(x0=y-y.spacing/2)

def g3(field):
    return field.dz(x0=z+z.spacing/2)

def g3_tilde(field):
    return field.dz(x0=z-z.spacing/2)

# Time update equation for quasi-p state variable p
p_update = \
    g1_tilde(b * (1 + 2 * e) * g1(p)) + \
    g2_tilde(b * (1 + 2 * e) * g2(p)) + \
    g3_tilde(b * (1 - f * eta**2) * g3(p)) + \
    g3_tilde(b * f * eta * sqrt(1 - eta**2) * g3(m)) - \
    (2 - t.spacing * wOverQ) * p + \
    (t.spacing * wOverQ - 1) * p.backward

# Time update equation for quasi-shear state variable m
# NOTE: unused for this issue, but m appears in the expression so I left this
m_update = \
    g1_tilde(b * (1 - f) * g1(m)) + \
    g2_tilde(b * (1 - f) * g2(m)) + \
    g3_tilde(b * (1 - f  + f * eta**2) * g3(m)) + \
    g3_tilde(b * f * eta * sqrt(1 - eta**2) * g3(p)) - \
    (2 - t.spacing * wOverQ) * p + \
    (t.spacing * wOverQ - 1) * p.backward

# Stencil for time updates
stencil_p = Eq(p.forward, p_update)

# Source
src = RickerSource(name='src', grid=v.grid, f0=fpeak, npoint=1, time_range=time_axis)
src.coordinates.data[:] = src_coords[:]
src_term = src.inject(field=p.forward, expr=src * t.spacing**2 * v**2 / b)

# Append the time discretization to the spacing map
dt = time_axis.step
spacing_map = v.grid.spacing_map
spacing_map.update({t.spacing: dt})
info(spacing_map)

# Build the operator
# Substitue scalar values for the space and time discretization
op = Operator([stencil_p, src_term], subs=spacing_map, name='OpExampleIso')
# op = Operator([stencil_p, src_term], subs=spacing_map, name='OpExampleIso', 
#               opt=('advanced', {'cire-repeats-inv': 2}))

op()
```

**Comment by FabioLuporini** [comment]
#1296 provides the mechanisms to fix this

```
def callback(n):
    # TODO: note that is currently naive. But it does the trick nicely
    # 50/20 are "min op counts to trigger a lifting"
    if n == 1:
        return 50
    elif n == 0:
        return 20
    assert False
op = Operator([stencil_p, src_term], subs=spacing_map, name='OpExampleIso', 
              opt=('advanced', {'cire-repeats-inv': 2, 'cire-mincost-inv': callback}))
```

**Comment by FabioLuporini** [comment]
I'm fairly sure this is now fixed by #1465, so closing. Feel free to reopen @jkwashbourne if something still feels off when u try it
